From ade73379860b5c768d10365e01b7d3bbf3a95cea Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Thu, 02 Mar 2023 10:50:00 +0200
Subject: [PATCH] [SW-125999] lib: string: add memcpy32_rev function

This patch adds memcpy32_rev function which copies
the data from destination to source in reverse order.
This function will be used by bootinfo for copying
to/from persistent memory to spsram.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I3fb615c3fdf5fc4207ceba95ce59b31f2eb53b1f
---

diff --git a/include/string.h b/include/string.h
index f4e61d5..b74e7ae 100644
--- a/include/string.h
+++ b/include/string.h
@@ -17,6 +17,7 @@
 void *memmove(void * dest,const void *src,size_t count);
 void *memset(void * s,int c,size_t count);
 void *memset32(uint32_t *s, uint32_t v, size_t count);
+void memcpy32_rev(void *dst, void *src, int size);
 
 int memcmp(const void * cs,const void * ct, size_t count);
 int strcmp(const char * cs,const char * ct);
diff --git a/lib/common/string.c b/lib/common/string.c
index 08dce41..a315bf0 100644
--- a/lib/common/string.c
+++ b/lib/common/string.c
@@ -220,3 +220,25 @@
 
 	return tmp;
 }
+
+/**
+ * memcpy32_rev - Copy one area of memory to another in reverse order
+ * @dest: Where to copy to
+ * @src: Where to copy from
+ * @count: The size of the area.
+ *
+ * You should not use this function to access IO space, use memcpy_toio()
+ * or memcpy_fromio() instead.
+ */
+void memcpy32_rev(void *dst, void *src, int size)
+{
+	uint32_t *dst32 = dst;
+	uint32_t *src32 = src;
+	int size32, i = 0;
+
+	for (size32 = size / 4; size32; size32--) {
+		dst32[size32 - 1] = src32[size32 - 1];
+		if (src32[size32 - 1] == 0)
+			i++;
+	}
+}
